### PR TITLE
Add OpenJDK 11, 12 and 13

### PR DESCRIPTION
**Pull request by reitermarkus**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

### CODE REVIEW

**Comment by reitermarkus** [comment]
@BrewTestBot test this please.

**Comment by fxcoudert** [comment]
Our usual policy for bootstrapping (see GCC, ghc, etc.) is to make the formula use an older binary (included as resource) for building. I don't think there is added value in maintaining a "chain of dependencies" on older versions of the same formula (on the contrary).

File: Formula/openjdk.rb [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+  depends_on "autoconf" => :build
+  depends_on "openjdk@12" => :build
+  depends_on "freetype"
+
+  def install
+    ENV["JAVA_HOME"] = Formula["openjdk@12"].libexec/"openjdk-12.jdk/Contents/Home"
+    java_options = ENV.delete("_JAVA_OPTIONS")
+
+    inreplace "make/autoconf/lib-freetype.m4",
```
Comment (fxcoudert) [inline Formula/openjdk.rb]: Has this been proposed upstream? Why is the reason they won't allow a non-bundled freetype?

File: Formula/openjdk.rb [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+                          "--with-boot-jdk=#{ENV["JAVA_HOME"]}",
+                          "--with-boot-jdk-jvmargs=#{java_options}",
+                          "--with-debug-level=release",
+                          "--with-native-debug-symbols=none",
+                          "--enable-dtrace=auto",
+                          "--with-jvm-variants=server",
+                          "--with-freetype-include=#{Formula["freetype"].include}",
+                          "--with-freetype-lib=#{Formula["freetype"].lib}"
+
+    ENV["MAKEFLAGS"] = ENV["MAKEFLAGS"]&.sub("-j", "JOBS=")
```
Comment (fxcoudert) [inline Formula/openjdk.rb]: Maybe this is simpler to read?

```suggestion
    ENV["MAKEFLAGS"] = "JOBS=#{ENV.make_jobs}"
```

File: Formula/openjdk.rb [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+  def caveats
+    <<~EOS
+      To use this JDK, symlink it with
+
+        sudo ln -sfn #{libexec}/openjdk-13.jdk /Library/Java/JavaVirtualMachines/
+
+      or set JAVA_HOME:
+
+        export JAVA_HOME="#{libexec}/openjdk-13.jdk/Contents/Home"
```
Comment (fxcoudert) [inline Formula/openjdk.rb]: How does this compare to what the cask currently does?

Comment (reitermarkus) [reply Formula/openjdk.rb]: The casks install directly to `/Library/Java/JavaVirtualMachines`.

File: Formula/openjdk.rb (hunk 1 above)
Comment (reitermarkus) [reply Formula/openjdk.rb]: Seems like the reason is “building freetype separately is a pain”.

https://bugs.openjdk.java.net/browse/JDK-8193017

File: Formula/openjdk.rb (hunk 3 above)
Comment (reitermarkus) [reply Formula/openjdk.rb]: Removed the `JAVA_HOME` part for now in any case to fix the audit error.

File: Formula/openjdk.rb (hunk 1 above)
Comment (reitermarkus) [reply Formula/openjdk.rb]: Removed this for now.

**Comment by reitermarkus** [comment]
@BrewTestBot test this please.

**Comment by fxcoudert** [comment]
@BrewTestBot test this please

**Comment by reitermarkus** [comment]
@BrewTestBot test this please.

**Comment by reitermarkus** [comment]
@BrewTestBot test this please

**Comment by reitermarkus** [comment]
@BrewTestBot test this please.

**Comment by reitermarkus** [comment]
@BrewTestBot test this please

**Comment by fxcoudert** [comment]
Bottling fails on Mojave and Catalina:

```
Error: Failed changing dylib ID of /usr/local/Cellar/openjdk/13.0.1+9/libexec/openjdk-13.jdk/Contents/Home/lib/libmlib_image.dylib
  from /usr/local/opt/openjdk/libexec/openjdk-13.jdk/Contents/Home/lib/libmlib_image.dylib
    to @@HOMEBREW_PREFIX@@/opt/openjdk/libexec/openjdk-13.jdk/Contents/Home/lib/libmlib_image.dylib
==> Bottling openjdk--13.0.1+9.mojave.bottle.tar.gz...
Error: Updated load commands do not fit in the header of /usr/local/Cellar/openjdk/13.0.1+9/libexec/openjdk-13.jdk/Contents/Home/lib/libmlib_image.dylib. /usr/local/Cellar/openjdk/13.0.1+9/libexec/openjdk-13.jdk/Contents/Home/lib/libmlib_image.dylib needs to be relinked, possibly with -headerpad or -headerpad_max_install_names
```

We need to make sure `-headerpad_max_install_names` is passed throughout the build.

**Comment by reitermarkus** [comment]
Yeah, I already force-pushed that change.

**Comment by reitermarkus** [comment]
Am I using `@BrewTestBot test this please` wrong or is it just unreliable?

**Comment by fxcoudert** [comment]
It's a bit unreliable, and also your job may be queued and might appear in the future :)

**Comment by reitermarkus** [comment]
@BrewTestBot test this please

**Comment by reitermarkus** [comment]
@BrewTestBot test this please

**Review by fxcoudert — APPROVED** [review]

**Comment by reitermarkus** [comment]
Should we add a symlink, e.g. `prefix/"java_home"`?

**Comment by fxcoudert** [comment]
This one warrants a wider discussion, I think. @Homebrew/core what do you think?

**Comment by chenrui333** [comment]
> Should we add a symlink, e.g. `prefix/"java_home"`?

Yeah, I think we should have JAVA_HOME to point to the installed formula. Does this also manage the PATH change as well?

**Comment by reitermarkus** [comment]
If `JAVA_HOME` is set, `/usr/bin/java` will use the `java` inside it.

**Comment by SMillerDev** [comment]
I'm not sure i understand what you mean. You want a special directory for Java in `/usr/local`?

**Comment by reitermarkus** [comment]
No, just a symlink in the formula's prefix, e.g. `Formula["openjdk"].prefix/"java_home"`, so that the path stays the same when `Formula["openjdk"].libexec/"openjdk-13.jdk"` becomes `Formula["openjdk"].libexec/"openjdk-14.jdk"`. Or we could just install all of them as `openjdk.jdk` internally.

**Comment by SMillerDev** [comment]
Having a non-versioned directory in libexec seems like a good idea to me. I'd prefer the last option personally since the java_home prefix seems confusing.

**Comment by chenrui333** [comment]
Just curious, how does users going to navigate between the different versions? Have to manually d the symlink part?

**Comment by reitermarkus** [comment]
Users either have to symlink it into `/Library/Java/JavaVirtualMachines` or 

```
export JAVA_HOME="$(brew --prefix openjdk)/libexec/openjdk.jdk/Contents/Home"
```

File: Formula/openjdk.rb [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+  def install
+    boot_jdk_dir = Pathname.pwd/"boot-jdk"
+    resource("boot-jdk").stage boot_jdk_dir
+    boot_jdk = boot_jdk_dir/"Contents/Home"
+    java_options = ENV.delete("_JAVA_OPTIONS")
+
+    short_version, _, build = version.to_s.rpartition("+")
+
+    chmod 0755, "configure"
```
Comment (MikeMcQuaid) [inline Formula/openjdk.rb]: `+x`?

File: Formula/openjdk.rb [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    system "make", "images"
+
+    libexec.install "build/macosx-x86_64-server-release/images/jdk-bundle/jdk-#{short_version}.jdk" => "openjdk.jdk"
+    bin.install_symlink Dir["#{libexec}/openjdk.jdk/Contents/Home/bin/*"]
+  end
+
+  def caveats
+    <<~EOS
+      To use this JDK, symlink it with
+
```
Comment (MikeMcQuaid) [inline Formula/openjdk.rb]: Kill the space

File: Formula/openjdk.rb [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    ENV["MAKEFLAGS"] = "JOBS=#{ENV.make_jobs}"
+    system "make", "images"
+
+    libexec.install "build/macosx-x86_64-server-release/images/jdk-bundle/jdk-#{short_version}.jdk" => "openjdk.jdk"
+    bin.install_symlink Dir["#{libexec}/openjdk.jdk/Contents/Home/bin/*"]
+  end
+
+  def caveats
+    <<~EOS
+      To use this JDK, symlink it with
```
Comment (MikeMcQuaid) [inline Formula/openjdk.rb]: Would be good to clarify what "use" means here; it looks like `javac` is functional without this, for example.

**Comment by MikeMcQuaid** [comment]
Very cool, nice work on this @reitermarkus. You realise you're signing up for doing Xorg after this, right 😂?

Would you see this as being something that could be used by `JavaRequirement` once it's up and running?

File: Formula/openjdk.rb (hunk 4 above)
Comment (reitermarkus) [reply Formula/openjdk.rb]: I didn't find any other formula using  `+x`.

Comment (MikeMcQuaid) [reply Formula/openjdk.rb]: Cool. Doesn't matter either way.

File: Formula/openjdk.rb (hunk 6 above)
Comment (reitermarkus) [reply Formula/openjdk.rb]: Which `javac`? This is needed for the system `java` wrapper to find it.

Comment (reitermarkus) [reply Formula/openjdk.rb]: `javac` will be functional if you have any JDK installed in `/Library/Java/JavaVirtualMachines`.

**Comment by reitermarkus** [comment]
@BrewTestBot test this please

File: Formula/openjdk.rb (hunk 6 above)
Comment (MikeMcQuaid) [reply Formula/openjdk.rb]: Cool, thanks. That helps and makes clearer it's optional.

**Comment by reitermarkus** [comment]
This should be ready to merge.

**Comment by sjackman** [comment]
Any comment on using `openjdk` vs `adoptopenjdk`?
Can we now use this formula instead of the cask as a dependency?

**Comment by reitermarkus** [comment]
> Any comment on using `openjdk` vs `adoptopenjdk`?

I don't think there is any significant difference between them. AdoptOpenJDK builds from upstream sources, so do we.

> Can we now use this formula instead of the cask as a dependency?

Yes.

**Comment by MikeMcQuaid** [comment]
> Can we now use this formula instead of the cask as a dependency?

Let's think a bit more strategically about how we do this. We don't want to just start swapping out random dependencies willy-nilly without thinking about the wider communication here. 2.2.0 will land soon so would be good to time communication with that.

**Comment by reitermarkus** [comment]
> how we do this

Here's how I think we should do this:

- If the formula just happens to be a tool written in Java, depend on `openjdk`.
- If the formula is a tool for Java development, depend on `openjdk`, but allow overriding `JAVA_HOME`.
- If the formula Depends on Java <= 1.8, keep `depends on :java`.

**Comment by MikeMcQuaid** [comment]
> * If the formula just happens to be a tool written in Java, depend on `openjdk`.

👍 

> * If the formula is a tool for Java development, depend on `openjdk`, but allow overriding `JAVA_HOME`.

Can you elaborate on this? Do you mean use one of the wrappers for this?



> * If the formula depends on Java <= 1.8, keep `depends on :java`.

👍 

I think we should hold off doing any of these until we've announced it and given a date in 2.2.0, though.

**Comment by reitermarkus** [comment]
> Do you mean use one of the wrappers for this?

What I did so far was use

```ruby
:JAVA_HOME => "${JAVA_HOME:-#{Formula["openjdk"].opt_prefix}}"
```

similar to `Language::Java::overridable_java_home_env`.

**Comment by joewiz** [comment]
I'm curious about this part of the plan:

> If the formula depends on Java <= 1.8, keep `depends on :java`.

Here's `java` as of now:

```
$ brew cask info java
java: 13.0.1,9:cec27d702aa74d5a8630c65ae61e4305
https://openjdk.java.net/
Not installed
From: https://github.com/Homebrew/homebrew-cask/blob/master/Casks/java.rb
==> Name
OpenJDK Java Development Kit
==> Artifacts
jdk-13.0.1.jdk -> /Library/Java/JavaVirtualMachines/openjdk-13.0.1.jdk (Generic Artifact)
```

How could this satisfy a formula that depends on Java <= 1.8? Will the `java` cask be changed to one of the OpenJDK flavors that supports Java 1.8?

**Comment by reitermarkus** [comment]
> How could this satisfy a formula that depends on Java <= 1.8?

For `depends_on :java => "1.8"` we are a currently recommending to install the `adoptopenjdk8` cask. This will stay the same.

File: Formula/openjdk.rb [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+  sha256 "97328e767bc5f47b097ec0e9d88a6a650e60c448dbaba2e835284a2bf5594eb5"
+
+  keg_only :provided_by_macos
+
+  depends_on "autoconf" => :build
+
+  resource "boot-jdk" do
+    url "https://download.java.net/java/GA/jdk12.0.2/e482c34c86bd4bf8b56c0b35558996b9/10/GPL/openjdk-12.0.2_osx-x64_bin.tar.gz"
+    sha256 "675a739ab89b28a8db89510f87cb2ec3206ec6662fb4b4996264c16c72cdd2a1"
+  end
```
Comment (bayandin) [inline Formula/openjdk.rb]: Shouldn't `boot-jdk` version be the same as the formula version?

Comment (reitermarkus) [reply Formula/openjdk.rb]: It's recommended building JDK n with JDK (n - 1).

Comment (bayandin) [reply Formula/openjdk.rb]: Thanks for the explanation!

**Pull request closed without merging**